index.js

```diff
@@ -63,41 +77,52 @@ function fastifyCors (fastify, opts, next) {
         return next(new Error('Invalid CORS origin option'))
       }
 
-      // Enable preflight
-      req.corsPreflightEnabled = true
-
-      reply.header('Access-Control-Allow-Origin',
-        getAccessControlAllowOriginHeader(req.headers.origin, resolvedOriginOption))
-
-      if (credentials) {
-        reply.header('Access-Control-Allow-Credentials', 'true')
-      }
-
-      if (exposedHeaders !== null) {
-        reply.header(
-          'Access-Control-Expose-Headers',
-          Array.isArray(exposedHeaders) ? exposedHeaders.join(', ') : exposedHeaders
-        )
+      addCorsHeaders(req, reply, resolvedOriginOption)
+
+      if (req.raw.method === 'OPTIONS' && preflight === true) {
+        req.corsPreflightEnabled = true
+
+        // Strict mode enforces the required headers for preflight
+        if (strictPreflight === true && (!req.headers.origin || !req.headers['access-control-request-method'])) {
+          reply.status(400).type('text/plain').send('Invalid Preflight Request')
+          return
+        }
+
+        addPreflightHeaders(req, reply)
+
+        if (!preflightContinue) {
+          // Do not call the hook callback and terminate the request
+          // Safari (and potentially other browsers) need content-length 0,
```

Comment:
Ain't this an issue in fastify itself?

Reply:
Not sure the history behind this. It is how it has always been implemented in fastify-cors.